src/backend/distributed/planner/merge_planner.c

```diff
@@ -13,92 +13,323 @@
+		CreateRouterMergePlan(targetRelationId, originalQuery, query,
+							  rangeTableList, plannerRestrictionContext);
+
+	/* Step 3: If the routing plan failed, try for repartition strategy */
+	if (distributedPlan->planningError != NULL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "For MERGE command, both the source and target "
-							 "must be distributed", NULL, NULL);
+		RaiseDeferredError(distributedPlan->planningError, DEBUG1);
+
+		/* If MERGE is not routable, try repartitioning  */
+		distributedPlan =
+			CreateRepartitionMergePlan(targetRelationId, planId,
+									   originalQuery, query,
+									   plannerRestrictionContext,
+									   boundParams);
 	}
 
-	/* All distributed tables must be colocated */
-	if (!AllDistributedRelationsInRTEListColocated(distTablesList))
+	return distributedPlan;
+
+	#endif
+}
+
+
+#if PG_VERSION_NUM >= PG_VERSION_15
+
+/*
+ * CreateRouterMergePlan attempts to create a pushable plan for the given MERGE
+ * SQL statement. If planning fails ->planningError is set to a description
+ * of the failure.
+ */
+static DistributedPlan *
+CreateRouterMergePlan(Oid targetRelationId, Query *originalQuery, Query *query,
+					  List *rangeTableList,
+					  PlannerRestrictionContext *plannerRestrictionContext)
+{
+	DistributedPlan *distributedPlan = CitusMakeNode(DistributedPlan);
+
+	Assert(originalQuery->commandType == CMD_MERGE);
+	Assert(OidIsValid(targetRelationId));
+
+	distributedPlan->planningError = DeferErrorIfRoutableMergeNotSupported(originalQuery,
+																		   rangeTableList,
+																		   plannerRestrictionContext);
+	if (distributedPlan->planningError != NULL)
 	{
-		return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-							 "For MERGE command, all the distributed tables "
-							 "must be colocated", NULL, NULL);
+		return distributedPlan;
 	}
 
-	return NULL;
+	Var *insertVar =
+		FetchAndValidateInsertVarIfExists(targetRelationId, originalQuery);
+	if (insertVar &&
+		!IsDistributionColumnInMergeSource((Expr *) insertVar, originalQuery, true))
+	{
+		ereport(ERROR, (errmsg("MERGE INSERT must use the source table "
+							   "distribution column value")));
+	}
+
+	Job *job = RouterJob(originalQuery, plannerRestrictionContext,
+						 &distributedPlan->planningError);
+
+	if (distributedPlan->planningError != NULL)
+	{
+		return distributedPlan;
+	}
+
+	ereport(DEBUG1, (errmsg("Creating MERGE router plan")));
+
+	distributedPlan->workerJob = job;
+	distributedPlan->targetRelationId = targetRelationId;
+	distributedPlan->modLevel = RowModifyLevelForQuery(query);
+
+	/* There is no coordinator query for MERGE */
+	distributedPlan->combineQuery = NULL;
+
+	/* MERGE doesn't support RETURNING clause */
+	distributedPlan->expectResults = false;
+	distributedPlan->fastPathRouterPlan =
+		plannerRestrictionContext->fastPathRestrictionContext->fastPathRouterQuery;
+
+	return distributedPlan;
+}
+
+
+/*
+ * CreateRepartitionMergePlan
```

Comment:
maybe update the comment, mention pull to coordinator piece as well

Reply:
Yup, completely missed the place-holder. Fixed it now.